dCommon/AMFDeserialize.cpp

```diff
@@ -0,0 +1,151 @@
+    inStream->Read(sizeOfString);
+
+    if (sizeOfString % 2 == 0) isReference = true;
+
+    sizeOfString = sizeOfString >> 1;
+
+    while (sizeOfString > 0) {
+        // Read the key in if this this is not a reference.  Otherwise get its reference.
+        std::string key;
+        if (!isReference) {
+            for (uint32_t i = 0; i < sizeOfString; i++) {
+                int8_t character;
+                inStream->Read(character);
+                key.push_back(character);
+            }
+            // Empty strings are never sent by reference.  This should never be empty here though...
+            if (!key.empty()) accessedElements.push_back(key);
+        } else {
+            key = accessedElements[sizeOfString];
+        }
+        // Read in the value type from the bitStream
+        int8_t valueType;
+        inStream->Read(valueType);
+        // Based on the typing, read a different value
+        switch (valueType) {
+            case AMFValueType::AMFUndefined: {
+                AMFUndefinedValue* undefinedValue = new AMFUndefinedValue();
+                values->InsertValue(key, undefinedValue);
+                break;
+            }
+
+            case AMFValueType::AMFNull: {
+                AMFNullValue* nullValue = new AMFNullValue();
+                values->InsertValue(key, nullValue);
+                break;
+            }
+
+            case AMFValueType::AMFFalse: {
+                AMFFalseValue* falseValue = new AMFFalseValue();
+                values->InsertValue(key, falseValue);
+                break;
+            }
+
+            case AMFValueType::AMFTrue: {
+                AMFTrueValue* trueValue = new AMFTrueValue();
+                values->InsertValue(key, trueValue);
+                break;
+            }
+
+            case AMFValueType::AMFDouble: {
+                AMFDoubleValue* doubleValue = new AMFDoubleValue();
+                double value;
+                inStream->Read(value);
+                doubleValue->SetDoubleValue(value);
+                values->InsertValue(key, doubleValue);
+                break;
+            }
+
+            case AMFValueType::AMFString: {
+                AMFStringValue* stringValue = new AMFStringValue();
+                std::string value;
+                int8_t length;
+                inStream->Read(length);
+                if (length % 2 == 1)  {
+                    length = length >> 1;
+                    for (uint32_t i = 0; i < length; i++) {
+                        int8_t character;
+                        inStream->Read(character);
+                        value.push_back(character);
+                    }
+                    // Empty strings are never sent by reference
+                    if (!value.empty()) accessedElements.push_back(value);
+                } else {
+                    // Length is a reference!
+                    length = length / 2;
```

Comment:
Should this for consistency with the other lines here also be `length >> 1` instead?